### ISSUE TITLE
Proposals for EmbedIO v3.0

### ISSUE THREAD

**Issue by rdeago** [unosquare/embedio#282]
It all started as an attempt to improve over recent changes to `StaticFilesModule`... but then, after some days working on EmbedIO, especially after @geoperez expressed his favour to breaking changes (as long as they actually improve the library), I noticed a couple of things that could be further simplified. Then another couple of things that needed some utility methods to avoid copy-pasting all over the place. Then a couple of places where data encapsulation needed some love. Then...

Today I looked at the work in progress that is my PR and... man. It is _huge_.

So I decided to keep track here of the changes I'm making, while I (hopefully) remember them all.

#### Guidelines ####

* Enforce the [Principle of least surprise](https://en.wikipedia.org/wiki/Principle_of_least_astonishment) wherever possible.

* Make it easier for other developers to peruse EmbedIO source code. _"It's an F12 keypress away"_ is no excuse when browsing GitHub.

* Make it easier for other develoeprs to write reusable modules.

* Do not expose APIs that should not be used (or abused) by external code.

* Expose APIs that can be useful to module developers.

* Features of base classes / interfaces that are only useful for a particular subclass / implementation should be moved to that subclass / implementation. Some examples are:

  * watchdogs, only actually used by `LocalSessionModule`;

  * web handlers and routing strategy, only actually used by `WebApiModule`.

* Make it possible to unit test a single module, or a group of modules, without a server. HTTP context and a `CancellationToken` should be enough for a module to do its work.

* Make test helper objects share as much code as possible with "real" objects.

* Remove features that are duplicates of .NET Standard 2.0 features, but were not present in previous versions. Some examples are:

  * `Strings.StandardCultureInfo` (replaceable with `CultureInfo.InvariantCulture`);

  * `Strings.StandardStringComparer` (replaceable with `StringComparer.InvariantCulture` and/or `StringComparison.InvariantCulture`);

  * several `System.Net` types.

#### What I'm doing ####

- [x] Switch to .NET Standard 2.0

- [x] Modify `StaticFilesModule` as detailed in this thread

- [x] Create a generic collection type to hold modules, WebSocket servers, etc.

- [x] Create a collection type for modules and use it

- [x] ~Create a collection type for WebSocket servers and use it~ _(superseded by new `WebSocketModule`)_

- [x] Replace `ISessionWebModule` with `ISessionManager` (not derived from `IWebModule`)

- [x] Create a base class for web servers, derive `WebServer` and `TestWebServer` from it

- [x] Speed up the creation of session IDs in `LocalSessionModule`

- [x] Extract ID creation from `LocalSessionModule` in an exported utility class (can be useful e.g. for other session managers)

- [x] Utility class for method argument validation

- [x] Add a unique string `Id` to `IHttpContext` so modules can use it in log (no need to create one more unique ID; just use `Request.RequestTraceIdentifier`)

- [x] Make `IHttpContext.Items` read-only to ensure consistency (see the discussion for #276)

- [x] Remove `Strings.StandardCultureInfo` (use `CultureInfo.InvariantCulture` instead)

- [x] Remove `Strings.StandardStringComparer` (use `StringComparer.InvariantCulture` / `StringComparison.InvariantCulture` instead, except where case-insensitiveness make sense, in which case use `StringComparer.InvariantCultureIgnoreCase` / `StringComparison.InvariantCultureIgnoreCase`)

- [x] Remove `Strings.CommaSplitChars` and `Strings.CookieSplitChars`: elements of `readonly` arrays are _not_ read-only

- [x] Add extension methods for `string` to ease splitting

- [x] Make some extension methods `public` (in separate classes) as they are apt to be useful to module and application developers

- [x] Remove routing strategy at a web server level; use a base URL path for each module instead

- [x] Add a `ModuleGroup` class that implements `IWebModule`, has its own collection of modules, and dispatches requests among them the same way a web server does

- [x] Move EmbedIO to its own namespace (ReSharper, of which I am a happy and proud user,  makes it a lot easier); rename namespace `Unosquare.Net` to `EmbedIO.Net`

- [x] Reorganize source code: folders == namespaces, one type per file

- [x] Move `internal` types to `Internal` namespace(s)

- [x] Move `using` directives outside of namespace declarations, to force full declaration of used namespaces

- [x] Reorder all `using` directives alphabetically, with `System` namespaces first (more kudos to ReSharper)

- [x] Use only Regex-based routing

- [x] Conceptually separate routing from controller-based handling

- [x] Refactor `WebApiModule` to use the new routing APIs

- [x] Fluent extensions to initialize practically everything

- [x] Improve XMLdoc comments

- [x] Revamp `FileModuleBase`: virtualize physical storage access, move all possible features from `StaticFilesModule` to the base class, so derived full-featured modules are easier to create (see [below](#issuecomment-488278943))

- [x] Use the new `FileModuleBase` to create a `ZipFileModule` _(this was done by @geoperez, who thus was the first author of an extension to EmbedIO v3 - way to go Geo!)_

#### What I'm NOT doing yet (but I'd like to) ####

* Make `TestHttpClient` as similar as possible to `System.Net.HttpClient` (requires copying code from .NET Core, as we'd need some `internal` classes and/or methods, e.g. for cookie parsing)

#### What I'm definitely not doing  ####

I abandoned the idea of initializing a `WebServer` with a collection initializer: it does not give enough flexibility. Instead, I'm putting fluent extensions on steroids.

#### So what? ####

I'll need ~a couple~ ~some more days~ two full months (as it turned out to be) to finish and test the PR.

I feel I may have stepped on some red line here, as I'm making a lot more changes than were discussed. @geoperez, please don't be afraid to tell me if it is the case and to reject any of the changes above.

_(Original post follows)_

---

### Background ###

`StaticFilesModule` can be used in different scenarios.

Some of them involve directories whose contents are practically immutable during the server application's lifetime (web applications, tipically) while some others involve directories to which files are added and/or deleted over time.

Some of them need a default document name and/or default extension; others do not need them, or may need them to be set differently.

Moreover, the same application may need to support different scenarios at the same time. Think a simple surveillance application, where we may have:
* a web application, whose HTML, CSS, etc will never change unless the whole application is updated;
* a directory where photos and/or videos are added each minute and deleted automatically after a certain amount of time;
* another directory where PDF reports are generated on-demand and made available for download for the next 10 minutes.

#### 1. Global configuration: one size does not always fit all ####
If at least one served directory is mutable, i.e. its contents are meant to change over time, `StaticFilesModule` must be initialized with the `cacheMappedPaths` parameter set to `false` (the default is `true` for compatibility with versions before 2.6.0); otherwise, the mapped paths cache would grow indefinitely, ultimately leading to memory exhaustion.

This has obvious ripercussions on performance for those directories that _do not_ change: static HTML, CSS, scripts, and in general what web applications are made of.

Besides, different virtual directories may need different default document names, different default extensions, different values for settings that do not even exist yet.

In other words, each virtual directory is, potentially, its own usage scenario.

Therefore I think that settings should not be global to `StaticFilesModule`, but specific to each virtual directory, maybe allowing for a global default set of settings to apply where not specified otherwise.

#### 2. What's so special about `/`? ####
`StaticFilesModule` has a `FileSystemPath` property that represents the local path corresponding to `/`.

Internally, though, `/` is a virtual path like any other, at least since version 2.6.0. It is treated differently just for compatibility with previous versions.

The fact that `/` is so "special" also leads to ambiguity in initialization. When using the
```C#
StaticFilesModule(
    string fileSystemPath,
    Dictionary<string, string> headers,
    Dictionary<string, string> additionalPaths,
    bool useDirectoryBrowser)
```
constructor, `/` need not appear in the `additionalPaths` dictionary, while when using
```C#
StaticFilesModule(Dictionary<string, string> paths)
```
`/` must be the _first_ key in the `paths` dictionary. It being in any other position triggers a bug that would deserve its own issue.

Moreover, not every application needs to serve files in `/`.

IMHO `/` should be initialized exactly like any other virtual path, and it should be optional to serve it with `StaticFilesModule`.

#### 3. Mutable configuration: do we really need it? ####
Virtual paths may be registered / unregistered at any time. Default extension and default document name may be changed at any time.

While this flexibility _may_ be useful in some situations, its negative impact is all but negligible:
* the mapped paths cache must keep track of which settings were involved in the resolution of each path, and selectively purge related items each time a setting is changed;
* virtual path management must use a reader / writer lock to ensure consistency, placing a performance burden on each request;
* the code needed to support the above points makes up for nearly half the code in virtual path management, making it more complicated, less maintainable, and exponentially more susceptible to bugs.

Making the settings immutable once `StaticFilesModule` is created would greatly simplify code and speed up path mapping. So, do we absolutely, positively need to change settings? Which kind of scenario do mutable settings support that cannot be supported otherwise?

### The solution ###
For problems 1 and 2 above there is no other solution but to modify `StaticFilesModule` and related classes, including their APIs.

Problem 2 may be solved in two ways:
1. either adding an immutable `StaticFilesOptions` class and a `StaticFilesOptionsBuilder` where options may be set before constructing the module;
2. or throwing `InvalidOperationException` upon any attempt to modify a setting when the web server is running.

### Why not? ###
Two words: breaking changes.

Being new to the EmbedIO community, I don't know how many people actually use `StaticFilesModule`, let alone which kinds of scenario the majority of them needs.

I think it would be nice to collect some thoughts here: maybe @geoperez can tag some known users of EmbedIO so they can read this and reply.

### Who's going to do it? ###
If there is enough consensus about the modifications to carry forward, I am willing to submit a PR in a few days.

**Comment by bufferUnderrun** [comment]
In my opinion, breaking changes are not a problem as long as they are properly documented in a release notes.

I've migrate my mains webapps to embedio and they all use the `StaticFilesModule`. My primary use case is to serve html, cs or js files. The `StaticFilesModule` settings are loaded at app start and never change during run (need a embedio restart if so). PDF (meaning data export) are generated on the fly using the `WebApiModule`.

**Comment by geoperez** [comment]
I'm open to breaking changes. Let's go ahead!

**Comment by rdeago** [comment]
Already on it.

In the meantime, may I ask for some guidance on updating the documentation? I tried checking out the `gh-pages` branch but it is not clear (to me, at least) what I should update and how.

_\<rant\>BTW, a separate branch for GitHub pages? Seriously, GitHub?!\</rant\>_

**Comment by geoperez** [comment]
LOL I like the separated branch for the website. But also a docs folder in the master branch can be used!

Anyway, for documentation let me explain to you how it goes:

* We use the `README` file for general information, "cookbook" and some samples for newbies.
* The GitHub page at https://unosquare.github.io/embedio/ is autogenerated each commit using DocFx. This tool will create static HTML pages from C# XML Documentation. So documenting in the source code is a good practice and you can also add examples.
* Finally, for BREAKING CHANGES I usually create a release note (https://github.com/unosquare/embedio/releases) and I mentioned the breaking change in the Nuget Release Notes.

Do you think we should add something else?

**Comment by rdeago** [comment]
> LOL I like the separated branch for the website. But also a docs folder in the master branch can be used!

A docs folder can be updated while writing code. I find writing docs a lighter task if done in parallel with coding. But that's just me, probably: I assume people at GitHub knew what they were doing when they gave the option to put docs in a branch.

> Do you think we should add something else?

A contributing guide for documentation, maybe? 😁
Seriously, the easier you make it for contributors to update the docs, the more you can be sure that they reflect latest changes.

Anyway, I'm trying to add good XML docs to next PR. I'll try to update Markdown docs too.

**Comment by geoperez** [comment]
We need a general contributing guide. I have it on my todo list.

**Comment by rdeago** [comment]
Just a quick, partial preview of what's coming:

```C#
new StaticFilesModule(new StaticFilesModuleOptions()

    // Just use shortcuts....
    .AddVirtualPath("/", VirtualPath.Map("/var/www/html").AsImmutableDirectory("index.html", ".html"))
    .AddVirtualPath("/download", VirtualPath.Map("/var/my/files").AsMutableBrowsableDirectory())

    // ...or just set the options you need...
    .AddVirtualPath("/graphs", VirtualPath.Map("/tmp/graphs").WithDefaultExtension(".jpg").WithDirectoryBrowser())

    // ...or use a shortcut and then customize it
    .AddVirtualPath("/zip", VirtualPath.Map("/var/my/archives").AsImmutableBrowsableDirectory(null, ".zip").WithoutGzip())
);
```
(Still working on `AddDefaultHeader`)

Almost everything can be configured on a per-virtual-path basis. Only RAM cache size and default headers have module scope.

There are no "default" options for virtual paths, not even `"index.html"` as default document. This keeps the surprise factor at a minimum, as the options object returned by `VirtualPath.Map` has all properties (except for the local path) set to their natural defaults.

----------

@geoperez, do you mind if I make some other changes while I'm at it? I'd like to:
1. Get completely rid of `ReadOnlyDictionary` anywhere.
2. Update the default list of MIME types from [MimeTypeMap sources](https://raw.githubusercontent.com/samuelneff/MimeTypeMap/master/src/MimeTypes/MimeTypeMap.cs) (it's the same list you took from StackOverflow, properly maintained in a Git repository) adding an appropriate notice in `LICENSE` (it's MIT-licensed so no legal issues should arise) and acomment stating the URL where and when it was retrieved.
3. Change the type of `StaticFilesModule.DefaultHeaders` to `WebHeaderCollection`, as a dictionary isn't exactly the best data model for web headers.

Feel free to cherry-pick. 😉

**Comment by geoperez** [comment]
I like what are you doing! Regarding the other changes, I like the idea. Please go ahead.

**Comment by rdeago** [comment]
BTW @geoperez do you plan to create a 2.8.0 or 3.0 version with these changes?

I'm asking because I have to mention the version number in a few places in XML docs.

**Comment by geoperez** [comment]
I think we probably move to 3.0.0 because we have many breaking changes.

**Comment by rdeago** [comment]
Oh, and (I'm probably getting annoying, I'm sorry) is .NET Standard 1.3 support _really_ needed?

* It requires quite a few workarounds (the sheer number of `#if NETSTANDARD1_3` and `#if !NETSTANDARD1_3` instances in the code screams for revenge.)
* It has no `WebHeaderCollection`, which would require exposing the one copied from Mono, then copying headers one by one in the response header because the header collections have different types if one is using Microsoft HTTP Listener.
* It requires keeping Mono support types around (`WebHeaderCollection` being just one among many) when `System.Net` types could be used even by Unosquare's `HTTPListener`.
* In general, .NET Standard v1 sucks big time... but that's just a matter of personal taste.

How about the following list of platforms?
* .NET Standard 2.0
* .NET Framework 4.7.1, so netstandard.dll is not required for .NET Framework applications
* maybe .NET Core 2.0 for the same reason as above (to be verified, though)

@geoperez have you got any statistics on EmbedIO use on different platforms?

**Comment by geoperez** [comment]
I don't have them. But I don't know what target framework is used by the Xamarin devs. Do you think of moving to NETStandard 1.4 or 1.5 will be easier?

**Comment by rdeago** [comment]
Unfortunately, System.Net stuff such as `WebHeaderCollection` was added in .NET Standard 2.0.

According to [Microsoft documentation](https://docs.microsoft.com/en-us/dotnet/standard/net-standard), raising the required .NET Standard version from 1.3 to 2.0 would mean raising the required Xamarin platform version thus:
* Xamarin.iOS: 10.0 -> 10.14
* Xamarin.Mac: 3.0 -> 3.8
* Xamarin.Android: 7.0 -> 8.0

Please feel free to correct me if I'm wrong, as I'm not a Xamarin developer, but:
* Apple is always pushing developers to use the latest SDK version anyway;
* as for Android, you may use Xamarin.Android 8.0 to target Android v5.0 if you wish.

Therefore .NET Standard 2.0 doesn't look like too much of a requirement, although someone who actually uses Xamarin could probably answer more sensibly than I.

**Comment by geoperez** [comment]
Hello @pvanloo, @madnik7, @erossini, @sonic1015, @deMichael, @pushqrdx 

Sorry for bothering you guys, can anyone help us with the previous question?

**Comment by madnik7** [comment]
Hi,
You have a long story and I cannot understand your real needs described in the question. What I think is that StaticFileModule should serve static files. Virtual maps are good, but if users need more than that, maybe they would better implement their own GET instead of a file. Still, I am not sure because it isn't my need yet.

But as a user here is my story and my needs of Static File are serving.
I am so rely on Static Files to server my SPA and some user contents but EmbedIO modules doesn't serve my needs because:
1) I am using external drive on android which does not provide File path access so I could not server user created contents on that drives with EmbedIO modules.
2) I need to serve my SPA files and I couldn't find any appreciated way to serve them with EmbedIO modules i. I didn't know how to server Asset directory with that and any portable file system.

What I did was implementing a new Portable Storage which is an abstract file/directory provider and implement another module based Unosquare.Labs.EmbedIO.Modules.FileModuleBase. I have also add a zip provider (via .net standard 2.0) so now I can easily server my SPA as zip file via the portable storage. I added a SAF (Androdi Storage Access Framework) provider (via Xamarin) and a FileProvider (via .net standard 2.0) to work with EmbedIO for Windows and Android devices.
Now, I simply server the SPA via a zip resource.

**Note**: In my approach the users can implement their own provider instead configuring virtual providers.
Here is my project (PortableStorage not the EmbedIO module) if you are interested (MIT license):
https://github.com/madnik7/PortableStorage

By the way, I think is .NET Standard 2.0 is ripe now and too much better than previous one. If I was in your team I would thinking about switching into it. Also, I think it is too soon for version 3.0.0 because it is still prevew and many important libraries still rely on version 2.0 can cause much conflicts.

**Comment by geoperez** [comment]
@madnik7 sorry for not explain me properly. The question is what target framework usually you use in Xamarin projects?

**Comment by madnik7** [comment]
Now 8 but it should be 9 soon:
The official documents say:
When you upload an APK, it needs to meet Google Play’s target API level requirements. Starting August 1, 2019, Google Play requires that new apps target at least Android 9.0 (API level 28), and that app updates target Android 9.0 from November 1, 2019. Until these dates, new apps and app updates must target at least Android 8.0 (API level 26).

https://developer.android.com/distribute/best-practices/develop/target-sdk

**Comment by geoperez** [comment]
But the Xamarin version is related to API level at Google? My question is more related to the .NET framework.

**Comment by madnik7** [comment]
Xamarin support .Net Standard 2.0 and I Have to support the target API to be used. But they are pretty up-to-date. Now the support Android 9 (PIE) too.

I hope I understand your question and what you look for

**Comment by madnik7** [comment]
Sorry for many grammar and typo error!

**Comment by geoperez** [comment]
Don't worry! Now I understand what is going on with Xamarin. Thanks!

**Comment by rdeago** [comment]
Hello @madnik7, thanks for sharing your thoughts!

> What I think is that StaticFileModule should serve static files.

You are touching an interesting point here: since `StaticFilesModule` serves both static and... not-so-static files, its name may be slightly misleading.

Actually, instead of users having to deal with configuring "immutable" vs. "mutable" mapped directories, a better way could be to have a `StaticFilesModule` _and_ a... `DynamicFilesModule` or something of the kind. The first with RAM cache and mapping cache, the second without them. @geoperez, how about it?

> Virtual maps are good, but if users need more than that, maybe they would better implement their own GET instead of a file.

Actually, virtual paths are there for serving files under different paths (think `Alias` in Apache). From what I've gathered by studying EmbedIO sources so far, virtual paths exist for two reasons:
* there is practically no support for having more than one module of the same type in the same `WebServer` (this IMHO needs to change, but that's beyond the point);
* even if we could simply define a `StaticFilesModule` for each URL path as needed, they wouldn't share the RAM cache, so it would be harder for a user to control their memory footprint.
It's not that these issues couldn't be addressed some other way, but this the situation as of now. I'll admit "virtual paths" may be another misleading name, though.

> I am so rely on Static Files to serve my SPA and some user contents, but EmbedIO modules don't serve my needs because:
> 
> 1. I am using external drive on android which does not provide File path access so I could not serve user created content on that drives with EmbedIO modules.
> 2. I need to serve my SPA files and I couldn't find any appreciated way to serve them with EmbedIO modules i. I didn't know how to serve Asset directory with that and any portable file system.
> 
> What I did was implementing a new Portable Storage which is an abstract file/directory provider and implement another module based Unosquare.Labs.EmbedIO.Modules.FileModuleBase. I have also added a zip provider (via .net standard 2.0) so now I can easily serve my SPA as zip file via the portable storage. I added a SAF (Android Storage Access Framework) provider (via Xamarin) and a FileProvider (via .net standard 2.0) to work with EmbedIO for Windows and Android devices.
>
> Now, I simply serve the SPA via a zip resource.

So, at least in Xamarin.Android, EmbedIO can do little more than serve a Web API, unless one jumps through the same hoops you did? That's good food for thought.

> Here is my project (PortableStorage not the EmbedIO module) if you are interested (MIT license):
> https://github.com/madnik7/PortableStorage

You really did an amazing job there. The issues you addressed are among those that frustrated me to no end in my (admittedly very limited and... not exactly pleasant) experience with Xamarin. I wish I came to know your library before!

> By the way, I think is .NET Standard 2.0 is ripe now and too much better than previous one. If I was in your team I would thinking about switching into it. Also, I think it is too soon for version 3.0.0 because it is still prevew and many important libraries still rely on version 2.0 can cause much conflicts.

I absolutely agree with you on both points.

Thanks again @madnik7 for your valuable input!

**Comment by rdeago** [comment]
So many thoughts, so little time. Let me try to sum things up:

### Xamarin support ###
Based on both @madnik7's input and some research of my own, requiring .NET Standard 2.0 doesn't seem to be a problem for Xamarin developers because:
* .NET Standard 2.0 support was introduced in Xamarin a year and a half ago (as per [Xamarin releases](https://developer.xamarin.com/releases/));
* one can use a newer Xamarin release to support older OS versions;
* one could not target too old an OS version anyway because of app store policies (both Apple's and Google's).

We also discovered that EmbedIO may have bigger problems with Xamarin than platform support: namely `StaticFilesModule` (and most probably `ResourceFilesModule`) being utterly useless on mobile platforms, because of limitations on filesystem access.

Not feeling qualified to solve this latter issue, I hope that someone with Xamarin experience helps us find a solution. [@madnik7's PortableStorage library](https://github.com/madnik7/PortableStorage) may be worth looking at in this regard.

### EmbedIO platform support ###
Looks like .NET Standard 2.0 is all we need to support. This is why .NET Standard exists, after all.

Maybe we could also support .NET Framework 4.7.1 and .NET Core 2.0 directly, so an application could potentially avoid loading `netstandard.dll` at all... but there are so many commonly-used .NET Standard 2.0 libraries out there, I wonder if this could actually make a difference.

### Serving static vs. non-static files ###
Splitting `StaticFilesModule` in two different modules, one with RAM cache and mapping cache and one without them, looks appealing to me.

It would mean two less boolean parameters on each virtual path, which would bring the number of arguments

Instead of this:
```C#
new StaticFilesModule(new StaticFilesModuleOptions()
    .AddVirtualPath("/", VirtualPath.Map("/var/www/html").AsImmutableDirectory("index.html", ".html"))
    .AddVirtualPath("/download", VirtualPath.Map("/var/my/files").AsMutableBrowsableDirectory())
    .AddVirtualPath("/graphs", VirtualPath.Map("/tmp/graphs").WithDefaultExtension(".jpg").WithDirectoryBrowser())
    .AddVirtualPath("/zip", VirtualPath.Map("/var/my/archives").AsImmutableBrowsableDirectory(null, ".zip").WithoutGzip())
);
```

we could have this:
```C#
// Serve would throw if the web server was already started at least once.
// Also, no "Unregister" / "Unserve".
new StaticFilesModule()
    .Serve("/", "/var/www/html", "index.html", ".html", false) // default for GZip is true
    .Serve("/zip", "/var/my/archives", null, ".zip", true, false));
new DynamicFilesModule() // needs a better name?
    .Serve("/download", "/var/my/files", null, null, true, true)
    .Serve("/graphs", "/tmp/graphs", null, ".jpg", true, false));

// Or call it Add instead of Serve and use collection initializers:
new StaticFilesModule {
    { "/", "/var/www/html", "index.html", ".html", false, true }, // Not browsable, use GZip
    { "/zip", "/var/my/archives", null, ".zip", true, false } // Browsable, no GZip
};
new DynamicFilesModule {
    { "/download", "/var/my/files", null, null, true, true },
    { "/graphs", "/tmp/graphs", null, ".jpg", true, false }
};
```
IMHO it looks a lot clearer, besides requiring a lot less code.

**Comment by madnik7** [comment]
Hi @rdeago
> We also discovered that EmbedIO may have bigger problems with Xamarin than platform support: namely StaticFilesModule (and most probably ResourceFilesModule) being utterly useless on mobile platforms, because of limitations on filesystem access.

.NET Standard 2.0 has already have File object so the supporting platform have to support it, It can be used for internal storage on both windows and android (don't know about iOS yet). The problem raises when we need to support external storage such as External Memory or USB OTG

> Looks like .NET Standard 2.0 is all we need to support. This is why .NET Standard exists, after all.

You should also considerate that it looks .NET Framework 4.7.2 is needed for full supported of .NET Standard 2.0 which is not already installed on Windows 7 or some versions of windows (Which personally I don't care but it may important for some people). .NET Standard 2.0 is a powerful and a big jump for multiplatform programming.

> Maybe we could also support .NET Framework 4.7.1 and .NET Core 2.0 directly, so an application could potentially avoid loading netstandard.dll at all... but there are so many commonly-used .NET Standard 2.0 libraries out there, I wonder if this could actually make a difference.

I cohldn't catch what you mean! .NET Standard just has a some abstration of system object which the target platform which support it has to implement it. Such as .NET Framework 4.7, .NET Core 2 and Xamarin that has implemented them. The rest is some system independent calculation and operation in SIL. Who cares about netstandard.dll. Even I am not sure what happen to it when a .Net Standard 2.0 library compiled by Xamarin.
I think embedIO can be implemented entirely as a .NET Standard 2.0 Library. So get rid of the the hassle of caring about different devices and platfroms.

**Comment by pvanloo** [comment]
> Hello @pvanloo, @madnik7, @erossini, @sonic1015, @deMichael, @pushqrdx
> 
> Sorry for bothering you guys, can anyone help us with the previous question?
> 
> EDIT: The question is: what target framework are you using in Xamarin projects? We are thinking about dropping NETStandard1.3 support soon.

Hi, FWIW I'm targeting standard 2.0

**Comment by rdeago** [comment]
Hi @madnik7

> .NET Standard 2.0 has already had the File object so the supporting platform has to support it, It can be used for internal storage on both windows and android (don't know about iOS yet). The problem arises when we need to support external storage such as External Memory or USB OTG

You said you had to write your own module to serve files through PortableStorage, not to mention having to write PortableStorage in the first place, to circumvent this limitation. I took it as an indication that _out-of-the-box EmbedIO_ is far from ideal as a HTTP server for mobile platforms, even if the problem is not caused by EmbedIO itself.

> You should also considerate that it looks .NET Framework 4.7.2 is needed for full supported of .NET Standard 2.0 which is not already installed on Windows 7 or some versions of windows (Which personally I don't care but it may important for some people).

Windows 7 [comes with .NET Framework 3.5.1 preinstalled](https://blogs.msdn.microsoft.com/astebner/2007/03/14/mailbag-what-version-of-the-net-framework-is-included-in-what-version-of-the-os/). I hope you'll agree that targeting it would hardly be worth the effort. Besides, .NET Framework versions up to and including v4.8 [can be installed](https://github.com/dotnet/docs/blob/master/docs/framework/get-started/system-requirements.md) on both Windows 7 and Windows Server (starting with v2008R2).

(To tell the truth, I didn't even bother to look at which versions can be installed on Vista, Windows 8, and Windows 8.1, as IMHO they will be sooner forgotten than MS forgiven for them.)

Targeting .NET Standard 2.0 leaves an application developer free to choose a version of .NET Framework still supported under Windows 7, namely v4.7.1, which according to [Microsoft docs](https://docs.microsoft.com/en-us/dotnet/framework/whats-new) was also the first version with _built-in_ support for .NET Standard 2.0. 

It even shipped with its own `netstandard.dll` instead of requiring developers to distribute one with applications.

> > Maybe we could also support .NET Framework 4.7.1 and .NET Core 2.0 directly [...]
> 
> I couldn't catch what you mean! [...] Who cares about netstandard.dll. Even I am not sure what happen to it when a .Net Standard 2.0 library compiled by Xamarin. [...]
> I think embedIO can be implemented entirely as a .NET Standard 2.0 Library. So get rid of the hassle of caring about different devices and platforms.

This has to do with how .NET Standard is implemented. I'll try to explain with an example.

Let's say EmbedIO targets .NET Standard 2.0 _only_, uses some nice .NET Standard goodies such as `System.ValueTuple` (a probable scenario in next version), and is used by an application targeting .NET Framework v4.6.1 or higher. Let's also pretend, for the sake of simplicity, that versions 4.6.1 through 4.7.1 of the framework _actually_ support .NET Standard 2.0 (even [Immo Landwerth had to admit](https://twitter.com/terrajobst/status/1031999730320986112) it was just wishful thinking.)

What happens when you build an application targeting, say, .NET Framework v4.6.1 and using EmbedIO?
* Since EmbedIO targets .NET Standard 2.0, _all type references_ in `EmbedIO.dll` point to `netstandard.dll`.
* `netstandard.dll` in turn contains the necessary information for the type loader to load the assembly which actually implements each type. This may be part of the framework (`System.Object` in `mscorlib.dll`, `System.Linq.Enumerable` in `System.Core.dll`, etc.) or not: v4.6.1 does not include `System.ValueTuple`, for example, so `System.ValueTuple.dll` would end up in ditribution too.
* NuGet knows which `netstandard.dll` to include in the application's project, according to the target platform of the application, along with other DLL's implementing types missing from the framework but required by the standard (`System.ValueTuple.dll` for example).
* Thus you end up with `Application.exe`, `EmbedIO.dll`, `netstandard.dll`, `System.ValueTuple.dll`, and possibly more DLL's being distributed.
* Besides, for each .NET Standard type referenced by `EmbedIO.dll`, the type loader has to "bounce" against `netstandard.dll` to find out which assembly to actually load.

Things change a bit if the application targets .NET Framework v4.7.1 or higher, as these have _built-in support_ for .NET Standard 2.0, meaning that `netstandard.dll`, `System.ValueTuple.dll`, and other parts that previously required separate NuGet packages are instead distributed with the framework. This means no `netstandard.dll` or `System.ValueTuple.dll` to distribute (NuGet and MSBuild take care of this), but the "bouncing" I mentioned above still happens. Admittedly, its actual impact on performance (especially upon application startup) may be negligible or not, depending on application demands, but it exists nonetheless.

What happens instead with a multi-targeted EmbedIO (`netstandard20` plus `net471`)?
* For applications targeting .NET Framework versions lower than v4.7.1, nothing, as they would load the .NET Standard version of `EmbedIO.dll`;
* For applications targeting .NET Framework version 4.7.1 or above, `EmbedIO.dll` would reference each type _directly in the assembly that actually contains it_. No bouncing, no loading `netstandard.dll` just for the bunch of metadata it contains.

Multi-targeting this way should require no changes to the code, because all APIs used by EmbedIO are available on both targets. So why not do it? Well, if the impact of using `netstandard.dll` is really negligible, it's of course better to stay on the side of simplicity and just target .NET Standard (cutting the NuGet package size in half, BTW). In the end, it's mainly a matter of how demanding we foresee applications to be in terms of startup performance.

In my previous post I also mentioned "other libraries" that could require `netstandard.dll` to be distributed, but it turns out I was wrong, as the requirement for its distribution depends upon the framework version targeted by the application.

What would, instead, be a _real_ hassle is multi-targeting .NET Framework versions before 4.7.1. Continuing with the example of `System.ValueTuple`, we should add a target-dependent dependency on the `System.ValueTuple` package to be able to build EmbedIO for .NET Framework 4.6.2, for example.

**Comment by rdeago** [comment]
> Hi, FWIW I'm targeting standard 2.0

Hi @pvanloo, thanks for your input. Looks like we could be on the right track by targeting `netstandard20`.

**Comment by geoperez** [comment]
@rdeago do you need to remove NETSTANDARD1.3 for your current branch? Or can we move this into another branch, and leaving only NETSTANDARD2.0?

After reading your comments, I agreed to use only one framework. Great explanation BTW 👍

**Comment by rdeago** [comment]
Since the commit story will be linear anyway, I could do everything in the PR I'm working on, if GitHub lets you merge a PR leaving the commits separate. Otherwise I'd prefer to do the platform switch first and then finish the files module(s), so I can use APIs not currently available (e.g. `ValueTuple`).

**Comment by geoperez** [comment]
I like the idea to merge the framework unification first. Can I help you with this PR today? Or do you already have it?

**Comment by rdeago** [comment]
Already on it. 😉

**Comment by geoperez** [comment]
Cool cool! Thanks.

**Comment by rdeago** [comment]
BTW, @geoperez, is the lack of `System.Net.HttpListener` in older platforms the only rationale for `Unosquare.Net.HttpListener`? If so, maybe it can be removed now. Anyway I'd leave this for another PR for now.

**Comment by geoperez** [comment]
I would like to keep Unosquare.Net.HttpListener for now.

**Comment by rdeago** [comment]
It's still there in #283. I stayed on the safe side.

**Comment by geoperez** [comment]
I just merged your PR #283. So you will continue from here to the original PR with the new StaticFilesModule, right? Do you need any help?

**Comment by rdeago** [comment]
I'll continue from there. Please tell me if you like the idea of two separate modules, with and without caching (see some posts above).

**Comment by geoperez** [comment]
> Actually, instead of users having to deal with configuring "immutable" vs. "mutable" mapped directories, a better way could be to have a StaticFilesModule and a... DynamicFilesModule or something of the kind. The first with RAM cache and mapping cache, the second without them.

I don't know if two modules should be the correct. This is because you want to keep the options set to the minimal or why? Maybe another strategy like a Factory?

**Comment by rdeago** [comment]
It's so that `StaticFilesModule` would deal with _really_ static files, while the other module (`UncachedFileModule`, maybe?) would deal with files that may change. Serving both with the same type of module seems somewhat confusing, as they have different purposes.

Once there are two distinct modules, thus removing the need for caching switches, options for each served path are few enough to allow for the simpler initialization code. This is a good thing, but it is not the reason why I proposed to split the module in the first place.

I don't get what the purpose of a Factory would be in this context (probably my fault, it's 11PM here 😪). A Factory for file-serving modules would make little sense, as a `WebServer` seems to support only one module per type. (Please remember that the same application could need to serve both cached and uncached, or immutable and mutable, directories at the same time.) On the other hand, a Factory for module configuration looks a little overengineered, doesn't it?

**Comment by madnik7** [comment]
Having an UncachedFileModule may not useful because sometimes there are files which frequently changed but need to be cached, in this case users havenothing to do but to restart the server module.

If the difference between the two module is just the cashing system, an option would better sense for me. By the way, personally I use little caching at lower level (StorageProvider) because the storage knows when changes happen and has more control it. Cashing small files contents would be little tricky especially finding the best time and place for releasing buffers, it has been always one of my most challenging experience.

**Comment by rdeago** [comment]
> Having an UncachedFileModule may not useful because sometimes there are files which frequently change but need to be cached

This is interesting. So far I admittedly haven't been able to think of a use case where files must be cached, yet they change frequently. What kind of files would you use it for?

> in this case, users have nothing to do but to restart the server module.

If EmbedIO compels users to restart a program, it's something that definitely needs fixing. Maybe we could limit the size of the mapping cache, at least for mutable directories that need caching.

> If the difference between the two module is just the cashing system, having an option would make better sense for me.

Absolutely agreed. Unfortunately, though, EmbedIO seems to have a "one module per type" policy: you can't have two `StaticFilesModule` instances in the same `WebServer`. So if you have to serve both a SPA (which you do via your own module on Android, but other people may do with `StaticFilesModule`) and files that may change, you're out of luck, unless:
* either caching is configurable separately for each path;
* or there is a separate module class for each caching scenario.

> By the way, personally I use little caching at lower level (StorageProvider) because the storage knows when changes happen and has more control it. Cashing small files contents would be little tricky especially finding the best time and place for releasing buffers, it has been always one of my most challenging experience.

That was exactly my rationale for not caching files that may change: I assume the underlying OS knows better, so the time and CPU saved by caching would be relatively little.

Besides, EmbedIO wouldn't know a file has changed until it is requested again. We could use a [`FileSystemWatcher`](https://docs.microsoft.com/en-us/dotnet/api/system.io.filesystemwatcher) to monitor served directories and free cache entries when files are modified or deleted, but in the end it would probably just make EmbedIO slower, rather than faster.

**Comment by geoperez** [comment]
> Absolutely agreed. Unfortunately, though, EmbedIO seems to have a "one module per type" policy: you can't have two StaticFilesModule instances in the same WebServer

And if we change this policy?

**Comment by rdeago** [comment]
Then we can have one `StaticFilesModule` for each directory (tree) to serve, each with its own parameters. Virtual paths would be no longer necessary. It remains to be decided if, and how, a `WebServer`-scoped RAM cache limit can be set, but overall it would drastically simplify the implementation.

**Comment by geoperez** [comment]
I was thinking about that, and there is not really a policy per se. This constraint is just because we need a way to lookup IWebModules, but using a named instances works fine too.

**Comment by rdeago** [comment]
OK, let me see if I can quickly do something about it.

**Comment by rdeago** [comment]
@geoperez, is it a problem if the `Name` property disappears from `IWebModule` and from the modules themselves, remaining only as a retrieval key in `WebModules`?

**Comment by geoperez** [comment]
I don't think so. Go ahead!

**Comment by rdeago** [comment]
Here's a little teaser: 😉 
```C#
public interface IWebModuleCollection : IReadOnlyList<IWebModule>, IReadOnlyDictionary<string, IWebModule>
{
}
```

**Comment by madnik7** [comment]
@rdeago 

> This is interesting. So far I admittedly haven't been able to think of a use case where files must be cached, yet they change frequently. What kind of files would you use it for?

It was just my imaginary situation:) But, we can imagine the home page of a website that contain news, also I think in such case people use an enterprise web server, so it should not be our case.

> So if you have to serve both a SPA (which you do via your own module on Android, but other people may do with StaticFilesModule) and files that may change

I really don't know how others can serve a SPA via StaticFilesModule on android! 

By the way, having multiple StaticFilesModules sounds good. It would be better it does not rely just on System File Path, I really had to read EmbedIO source code in depth and did dirty thing to create a class driven from FileModuleBase to change its behavior to force Emebed IO using the PortableStorage.

**Comment by rdeago** [comment]
Hi @madnik7
 
> It was just my imaginary situation:) But, we can imagine the home page of a website that contain news, also I think in such case people use an enterprise web server, so it should not be our case.

Anyway, for cases where files change frequently yet caching may help, I'm leaving an intermediate choice between "cache" and "no cache": that is, caching the mapping between URL and local file / directory and let the OS cache the files themselves.

> I really don't know how others can serve a SPA via StaticFilesModule on android!

Me neither. I think it'd be interesting to see how different have responded to this challenge.

> By the way, having multiple StaticFilesModules sounds good. It would be better it does not rely just on System File Path, I really had to read EmbedIO source code in depth and did dirty thing to create a class driven from FileModuleBase to change its behavior to force Emebed IO using the PortableStorage.

If there was a base class taking care of caching, eTags, etc. and you could derive from that, would that do the trick? It may be a good thing to do, because `ResourceFilesModule` could derive from it too; plus it would make a `ZipFileModule` quite trivial to develop.

**Comment by rdeago** [comment]
@geoperez @bufferUnderrun @madnik7 @pvanloo and practically everyone using EmbedIO: sorry to bother you, I need some guidance at this point.

1. @geoperez: is there any particular reason why you used [`RequestUri.LocalPath`](https://docs.microsoft.com/en-us/dotnet/api/system.uri.localpath) instead of [`RequestUri.AbsolutePath`](https://docs.microsoft.com/en-us/dotnet/api/system.uri.absolutepath)?

2. Anyone: is there any existing use case that requires case-insensitive URL paths in EmbedIO?

3. Last but not least: @geoperez, are you OK with the plan I've layed out in the edited first post? Anything to add / remove / modify?

**Comment by rdeago** [comment]
Unfortunately this time I didn't do separate commits. Too much work, too little time. That's why I didn't commit anything yet: in this very moment, the thing doesn't even compile!

I can commit on my fork tomorrow morning if you want. Mind you, the code isn't finished, but the core classes are 99% done.

Areas of code that need your help include:

* reducing the number of classes copied from Mono, now that we use .NET Standard 2.0 with its richer `System.Net`

* assistance with WebSockets (we should find a more direct way to collaborate on that than comments on this post)

* documentation: since there is going to be a v3.0, the docs for v2.0 should probably be copied over and kept as reference with a link in README, just as it is for v1.

* ...and this is only the beginning. Once you see the new code you'll probably find out more things to do.

**Comment by geoperez** [comment]
@rdeago If you wish I can create a branch in this repo, yo merge your changes there, and we can collaborate from there. I just need to add you as a contributor.

**Comment by rdeago** [comment]
> @rdeago If you wish I can create a branch in this repo, yo merge your changes there, and we can collaborate from there. I just need to add you as a contributor.

That would be fantastic! Thanks a lot.

**Comment by geoperez** [comment]
I just created the branch `v3.0`

**Comment by pvanloo** [comment]
I can see a use-case where case-sensitive ID's get used in a path (I wouldn't consider it good practice but still)

**Comment by rdeago** [comment]
@geoperez: you're going to have a busy day. 😁 

@pvanloo: thanks for your input.

Regarding case-sensitivity of URL paths, [RFC3986](https://tools.ietf.org/html/rfc3986#section-3.3) doesn't state whether they should be case-sensitive or not.

In "ancient" times, URL paths were more or less 1-to-1 mapped to either static files or CGI scripts; therefore, their case-sensitivity tended to be the same as that of the underlying file system. IIS treated URL paths as case-insensitive just like Windows did, while all other major Web servers ran under Unix-like operating systems and as a consequence treated paths as case-sensitive.

Thanks to .NET Core and Mono, we don't even know under which platform EmbedIO will actually run. OS detection would not help much, either: nowadays Linux and MacOS/X can also use NTFS, which is case-insensitive, right? No, it is not: _Windows_ is. Oh wait, [not any more](https://www.howtogeek.com/354220/how-to-enable-case-sensitive-folders-on-windows-10/): case-sensitivity can now vary on a folder-by-folder basis. 😱 

By the way, what does case-insensitivity even mean? Enter the Turkish variant of the Latin alphabet, where uppercase `i` is `İ` and lowercase `I` is `ı`. Clearly `CultureInfo.InvariantCulture` does not cut it, so should we use `CurrentCulture`, or `CurrentUICulture`? What if some guy from Turkey (or Azerbaijan, for that matter) uploads a file whose name contains a dotted `İ` in a directory I'm serving with `StaticFilesModule` on a Windows system in Italy (where uppercase `i` is `I`)? Should they expect to be able to retrieve that file using an all-lowercase path? To make a long story short, case-insensitivity is a big, ugly mess.

Besides, as @pvanloo pointed out (and RFC3986 clearly states) URL paths may contain data, which _may_ be case-sensitive: IDs, base64-encoded data, you name it.

All this considered, I'd say that, at least for now, we'd better treat URL paths as case-sensitive. This will cause files to be cached more than once by `StaticFilesModule` if requested with different casing. Personally I can live with it, but we may add case-insensitivity as an option at a module level in a later version if necessary.

---

Some additional reference links:

* [Why are URLs case-sensitive?](https://webmasters.stackexchange.com/questions/90339/why-are-urls-case-sensitive)

* [Should URL paths be case sensitive?](https://webmasters.stackexchange.com/questions/69705/should-url-paths-be-case-sensitive)

**Comment by rdeago** [comment]
All the described features (plus others added along the way) have been implemented.

EmbedIO v3.0 has been in preview on NuGet for some days.

I'm closing this issue. If there is any problem with the new version, it is more appropriate to open a new issue, as this one is already taking too long to load in my browser. 😄
